wallet/src/store/utxo.rs

```diff
@@ -82,40 +90,70 @@ impl<KEY> UtxoGroup<KEY> {
 
     /// utxos already ordered by value
     pub fn utxos(&self) -> impl Iterator<Item = &UTxO> {
-        self.by_value.values().flatten()
+        self.by_value
+            .iter()
+            .sorted_by_key(|x| x.0)
+            .map(|(_k, set)| set.iter())
+            .flatten()
     }
 
     /// total value of the given group
     pub fn total_value(&self) -> Value {
         self.total_value
     }
 
-    fn add(&mut self, utxo: UTxO) {
-        self.total_value = self.total_value.saturating_add(utxo.value);
-        self.by_value.entry(utxo.value).or_default().insert(utxo);
+    fn add(&self, utxo: UTxO) -> Self {
+        let Self {
+            total_value,
+            mut by_value,
+            key,
+        } = self.clone();
+
+        let total_value = total_value.saturating_add(utxo.value);
+
+        by_value = by_value.insert_or_update_simple(
+            utxo.value,
+            {
+                let set = HashSet::new();
+                let set = set.insert(utxo.clone());
+                set
+            },
+            |old| Some(old.insert(utxo.clone())),
+        );
+
+        Self {
+            total_value,
+            by_value,
+            key,
+        }
     }
 
-    fn remove(&mut self, utxo: &UtxoPointer) -> Option<UTxO> {
-        use im_rc::ordmap::Entry::*;
+    fn remove(&self, utxo: &UtxoPointer) -> Self {
+        // let mut new = self.clone();
 
-        if let Occupied(mut occupied) = self.by_value.entry(utxo.value) {
-            if let Some(prev) = occupied.get_mut().remove(utxo) {
-                self.total_value = self
-                    .total_value
-                    .checked_sub(prev.value)
+        let Self {
+            by_value,
+            mut total_value,
+            key,
+        } = self.clone();
+
+        let by_value = by_value
+            .update::<_, std::convert::Infallible>(&utxo.value, |set| {
+                let new_set = set.remove(&Rc::new(utxo.clone()));
```

Comment:
Is the `Rc::new` wrapper made unnecessary with https://github.com/input-output-hk/chain-libs/pull/474?

Reply:
well, not this one, because I only implemented it for the `lookup` function. But as that one got merged I assume it should be fine to do the same for the remove.

Reply:
Please extend that in imhamt, that API should be consistent.

Reply:
I just did in https://github.com/input-output-hk/chain-libs/pull/475

Reply:
I updated the code to that branch, but I can't update the submodule without handling the new private voting types, which I did in #107 . It's probably easier to merge that one first.

Reply:
#107 is merged now, as is nput-output-hk/chain-libs#475, please update.

Reply:
oh, I somehow missed that notification. I rebased and updated the submodule, I'll merge as soon as the ci turns green.